pinot-core/src/main/java/com/linkedin/pinot/core/common/datatable/DataTableBuilder.java

```diff
@@ -0,0 +1,316 @@
+ * Integer etc. This is waste of cpu resource and increase the payload size. We
+ * optimize the data format for Pinot usecase. We can also support lazy
+ * construction of obejcts. Infact we retain the bytes as it is and will be able
+ * to lookup the a field directly within a byte buffer.<br>
+ *
+ * USAGE:
+ *
+ * Datatable is initialized with the schema of the table. Schema describes the
+ * columnnames, their order and data type for each column.<br>
+ * Each row must follow the same convention. We don't support MultiValue columns
+ * for now. Format,
+ * |VERSION,DATA_START_OFFSET,DICTIONARY_START_OFFSET,INDEX_START_OFFSET
+ * ,METADATA_START_OFFSET | |&lt;DATA&gt; |
+ *
+ * |&lt;DICTIONARY&gt;|
+ *
+ *
+ * |&lt;METADATA&gt;| Data contains the actual values written by the application We
+ * first write the entire data in its raw byte format. For example if you data
+ * type is Int, it will write 4 bytes. For most data types that are fixed width,
+ * we just write the raw data. For special cases like String, we create a
+ * dictionary. Dictionary will be never exposed to the user. All conversions
+ * will be done internally. In future, we might decide dynamically if dictionary
+ * creation is needed, for now we will always create dictionaries for string
+ * columns. During deserialization we will always load the dictionary
+ * first.Overall having dictionary allow us to convert data table into a fixed
+ * width matrix and thus allowing look up and easy traversal.
+ *
+ *
+ */
+// TODO: potential optimizations:
+// TODO:   1. Fix float size.
+// TODO:   2. Use one dictionary for all columns (save space).
+// TODO:   3. Given a data schema, write all values one by one instead of using rowId and colId to position (save time).
+public class DataTableBuilder {
+  private final DataSchema _dataSchema;
+  private final int[] _columnOffsets;
+  private final int _rowSizeInBytes;
+  private final Map<String, Map<String, Integer>> _dictionaryMap = new HashMap<>();
+  private final Map<String, Map<Integer, String>> _reverseDictionaryMap = new HashMap<>();
+  private final ByteArrayOutputStream _fixedSizeDataByteArrayOutputStream = new ByteArrayOutputStream();
+  private final ByteArrayOutputStream _variableSizeDataByteArrayOutputStream = new ByteArrayOutputStream();
+  private final DataOutputStream _variableSizeDataOutputStream =
+      new DataOutputStream(_variableSizeDataByteArrayOutputStream);
+
+  private int _numRows;
+  private ByteBuffer _currentRowDataByteBuffer;
+
+  public DataTableBuilder(@Nonnull DataSchema dataSchema) {
+    _dataSchema = dataSchema;
+    int numColumns = dataSchema.size();
+    _columnOffsets = new int[numColumns];
+    int currentOffset = 0;
+    for (int i = 0; i < numColumns; i++) {
+      DataType dataType = dataSchema.getColumnType(i);
+      _columnOffsets[i] = currentOffset;
+      switch (dataType) {
+        case BOOLEAN:
+          currentOffset += 1;
```

Comment:
Currently V1Constants does not have size for several types including BOOLEAN, BYTE etc. Also, this is protocol related, so I will keep the magic numbers for now.